### DISCUSSION TITLE
how can I create a named-key array type?

### DISCUSSION THREAD

**Discussion by kkmuffme** [vimeo/psalm#8337] [Plugin development]
I know for `array<non-empty-string, int|string|list<string>>` I can do:

```
$type = new TArray(
    [
        Type::getNonEmptyString(),
        new Union([new TInt(), new TString(), new TNonEmptyList(Type::getString())]),
    ]
);

return new Union([$type]);
```

However I wasn't able to find an example for when I have this:

```
array{name: string|list<string>, type: int}
```

Could someone quickly guide me/point towards an example?

**Comment by AndrolGenhald** [comment]
https://psalm.dev/docs/running_psalm/plugins/plugins_type_system/#arrays

You want `TKeyedArray`.
```php
new TKeyedArray([
    "name" => new Union([
        new TString(),
        new TList(new Union([new TString()])),
    ]),
    "type" => new Union([new TInt()]),
]);
```

**Reply by kkmuffme** [reply]
Thanks

**Reply by kkmuffme** [reply]
Is there any way I could declare an array where some keys are named but there is an unknown number of unnamed keys with type int?

**Reply by AndrolGenhald** [reply]
Set `TKeyedArray::$previous_key_type` and `::$previous_value_type`. Note that this might change if I ever get around to adding support for index signature like arrays. I'd also advise against setting one of those but not the other, it _probably_ won't crash, but I'm not really sure.

**Reply by kkmuffme** [reply]
Could you clarify this with an example, bc I'm not sure I get that

**Reply by AndrolGenhald** [reply]
```php
$array_type = new TKeyedArray(...);
$array_type->previous_key_type = new Union(...);
$array_type->previous_value_type = new Union(...);
```
